junit-vintage-engine/src/main/java/org/junit/vintage/engine/descriptor/TestSourceProvider.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.vintage.engine.descriptor;
+
+import static java.util.function.Predicate.isEqual;
+import static java.util.stream.Collectors.toList;
+import static org.apiguardian.api.API.Status.INTERNAL;
+import static org.junit.platform.commons.util.FunctionUtils.where;
+import static org.junit.platform.commons.util.ReflectionUtils.findMethods;
+
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apiguardian.api.API;
+import org.junit.platform.commons.support.ModifierSupport;
+import org.junit.platform.commons.util.LruCache;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.junit.runner.Description;
+
+/**
+ * @since 5.6
+ */
+@API(status = INTERNAL, since = "5.6")
+public class TestSourceProvider {
+
+	private final Map<Class<?>, List<Method>> methodsCache = Collections.synchronizedMap(new LruCache<>(32));
+
+	public TestSource findTestSource(Description description) {
+		Class<?> testClass = description.getTestClass();
+		if (testClass != null) {
+			String methodName = description.getMethodName();
+			if (methodName != null) {
+				Method method = findMethod(testClass, methodName);
+				if (method != null) {
+					return MethodSource.from(testClass, method);
+				}
+			}
+			return ClassSource.from(testClass);
+		}
+		return null;
+	}
+
+	private Method findMethod(Class<?> testClass, String methodName) {
+		if (methodName.contains("[") && methodName.endsWith("]")) {
+			// special case for parameterized tests
+			return findMethod(testClass, methodName.substring(0, methodName.indexOf("[")));
```

Comment:
Why perform recursion here?

Why not just `methodName = methodName.substring(0, methodName.indexOf("["));`?

Reply:
Probably because I don't like reassigning method parameters. 🤔

Reply:
Hmmmm... hmmmmmmmmmm....

Reply:
If you really dislike reassigning method parameters, you could introduce a local variable such as `effectiveMethodName` to avoid the recursion, but in the end it's not a big deal for this use case.

Reply:
I extracted the first part into a separate method.

Reply:
That looks much better. Thanks.